canonical-web-and-design/maas-ui
# feat(ui): add custom certificate fields in LXD machine config

## Done

- Created `CertificateDetails` component which shows read-only info of an existing certificate, used in both machine and KVM details
- Implemented custom certificate fields in LXD machine config - if the user is not editing the form they see the certificate data, if the user is editing the form they see the radios for generating/updating a certificate
- Changed `AuthenticationFields` a bit. I've renamed it to `CertificateFields` which I think makes a bit more sense, but also I've removed the password field entirely from it. This makes it so I can reuse the component in exactly the same way in 4 different places, otherwise there's a bit of bespoke field placement for each case:

**Add machine**
![2021-09-24_17-21](https://user-images.githubusercontent.com/25733845/134635366-7f96d5c2-a7bb-4d71-91c9-1183e81f939c.png)
**Edit machine**
![2021-09-24_17-22](https://user-images.githubusercontent.com/25733845/134635334-817221c7-bc98-4192-a2d7-54eaf18b36e7.png)
**Add KVM**
![2021-09-24_17-22_1](https://user-images.githubusercontent.com/25733845/134635350-3d7d188c-cf78-4c6d-99bd-06c0a02eb884.png)
**Edit KVM**
![2021-09-24_17-23](https://user-images.githubusercontent.com/25733845/134635356-c0da3b24-daab-4946-93f2-2cd6d6e652a7.png)
- @amylily1011 the main consequence here is that in the "Add machine" form the password no longer shows below the "Generate new certificate" radio if its selected. Let me know if this is a deal breaker.

### QA steps

- Go to the config page of a LXD machine that was added manually, i.e. not composed from a VM host. Add a new one if it doesn't exist.
- Check that certificate data shows in the power section.
- Click "Edit" for the power config form and check that the certificate data changes into certificate fields.
- Select "Generate new certificate" and submit the form.
- Refresh the page (there's a bug in the backend where no machine update notify message is sent when updating machine power parameters).
- Check that the certificate updated successfully.
- Click "Edit" and try uploading a certificate/key manually.
- Refresh the page and check that the certificate updated successfully.
- Go to the config page of a LXD machine that was composed from a VM host. Compose a new one if it doesn't exist.
- Check that certificate data shows in the power section.
- Click "Edit" and check that you do not get the option to edit the certificate, since that is handled by the LXD VM host.

## Fixes

Fixes canonical-web-and-design/app-squad#247

## Launchpad issue

[lp#1944847](https://bugs.launchpad.net/maas/+bug/1944847)

## Screenshots
### Added LXD machine
![Peek 2021-09-24 18-11](https://user-images.githubusercontent.com/25733845/134641405-2fddf5b5-2597-4d6c-8e75-671864967b9f.gif)


### Composed LXD machine
![Peek 2021-09-24 18-12](https://user-images.githubusercontent.com/25733845/134641425-0c7c68e5-e0d2-443f-ab5e-677e8577ba64.gif)
